plugins/demoproxy.c

```diff
@@ -0,0 +1,218 @@
+ * make demoproxy.so
+ *
+ * Then copy or symlink the plugins/demoproxy.so and plugins/demoproxytest.px files
+ * to ~/.config/geany/plugins
+ * - it will be loaded at next startup.
+ */
+
+/* plugin API, always comes first */
+#include "geanyplugin.h"
+
+typedef struct {
+	GKeyFile       *file;
+	GPtrArray      *strings;
+	GSList         *menu_items;
+}
+PluginContext;
+
+static gboolean proxy_init(GeanyPlugin *plugin, gpointer pdata)
+{
+	PluginContext *data;
+	GKeyFile *file;
+	gchar fmt[] = "item%d";
+	gint i = 0;
+	gchar *text;
+	GeanyData *geany = plugin->geany_data;
+
+	data = (PluginContext *) pdata;
+	file = data->file;
+
+	/* Normally, you would instruct the VM/interpreter to call into the actual plugin. The
+	 * plugin would be identified by pdata. Because there is no interpreter for
+	 * .ini files we do it inline, as this is just a demo */
+	text = g_key_file_get_locale_string(data->file, "Help", "text", NULL, NULL);
+	data->strings = g_ptr_array_new();
+	g_ptr_array_set_free_func(data->strings, g_free);
+
+	g_ptr_array_add(data->strings, text);
+	while (TRUE)
+	{
+		GtkWidget *item;
+		gchar *key = g_strdup_printf(fmt, i++);
+		text = g_key_file_get_locale_string(file, "Init", key, NULL, NULL);
+		g_free(key);
+
+		if (!text)
+			break;
+
+		item = gtk_menu_item_new_with_label(text);
+		gtk_widget_show(item);
+		gtk_container_add(GTK_CONTAINER(geany->main_widgets->tools_menu), item);
+		gtk_widget_set_sensitive(item, FALSE);
+		data->menu_items = g_slist_prepend(data->menu_items, (gpointer) item);
+		g_ptr_array_add(data->strings, text);
+	}
+
+	return TRUE;
+}
+
+
+static void proxy_help(GeanyPlugin *plugin, gpointer pdata)
+{
+	PluginContext *data;
+	GtkWidget *dialog;
+	gchar *help_text;
+	GKeyFile *file;
+
+	data = (PluginContext *) pdata;
+	help_text = g_ptr_array_index(data->strings, 0);
+
+	dialog = gtk_message_dialog_new(
+		GTK_WINDOW(plugin->geany_data->main_widgets->window),
+		GTK_DIALOG_DESTROY_WITH_PARENT,
+		GTK_MESSAGE_INFO,
+		GTK_BUTTONS_OK,
+		"%s", help_text);
+	gtk_message_dialog_format_secondary_text(GTK_MESSAGE_DIALOG(dialog),
+		_("(From the %s plugin)"), plugin->info->name);
+
+	gtk_dialog_run(GTK_DIALOG(dialog));
+	g_free(help_text);
+	gtk_widget_destroy(dialog);
+}
+
+
+static void proxy_cleanup(GeanyPlugin *plugin, gpointer pdata)
+{
+	PluginContext *data;
+
+	data = (PluginContext *) pdata;
+
+	g_return_if_fail(data);
+	g_return_if_fail(data->menu_items);
+
+	g_slist_free_full(data->menu_items, (GDestroyNotify) gtk_widget_destroy);
+	g_ptr_array_unref(data->strings);
+}
+
+
+static gint demoproxy_probe(GeanyPlugin *proxy, const gchar *filename, gpointer pdata)
+{
+	/* We know the extension is right (Geany checks that). For demo purposes we perform an
+	 * additional check. This is not necessary when the extension is unique enough. */
+	gboolean match;
+	gchar linebuf[128];
+	FILE *f = fopen(filename, "r");
+	fgets(linebuf, sizeof(linebuf), f);
+	match = utils_str_equal(linebuf, "#!!proxy_MAGIC!!\n");
```

Comment:
the magic doesn't match the one in the docs or the one actually used in the sample *.px*